tensorboard/uploader/server_info.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Initial server communication to determine session parameters."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from google.protobuf import message
+import requests
+
+from tensorboard import version
+from tensorboard.uploader.proto import server_info_pb2
+
+
+# Request timeout for communicating with remote server.
+_REQUEST_TIMEOUT_SECONDS = 10
+
+
+def _server_info_request():
+  request = server_info_pb2.ServerInfoRequest()
+  request.version = version.VERSION
+  return request
+
+
+def fetch_server_info(origin):
+  """Fetches server info from a remote server.
+
+  Args:
+    origin: The server with which to communicate. Should be a string
+      like "https://tensorboard.dev", including protocol, host, and (if
+      needed) port.
+
+  Returns:
+    A `server_info_pb2.ServerInfoResponse` message.
+
+  Raises:
+    CommunicationError: Upon failure to connect to or successfully
+      communicate with the remote server.
+  """
+  endpoint = "%s/api/uploader" % origin
+  post_body = _server_info_request().SerializeToString()
+  try:
+    response = requests.post(
+        endpoint, data=post_body, timeout=_REQUEST_TIMEOUT_SECONDS
+    )
+  except requests.RequestException as e:
+    raise CommunicationError("Failed to connect to backend: %s" % e)
+  if not response.ok:
+    raise CommunicationError(
+        "Non-OK status from backend (%d %s): %r"
+        % (response.status_code, response.reason, response.content)
+    )
+  try:
+    return server_info_pb2.ServerInfoResponse.FromString(response.content)
+  except message.DecodeError as e:
+    raise CommunicationError(
+        "Corrupt response from backend (%s): %r" % (e, response.content)
+    )
+
+
+def create_server_info(frontend_origin, api_endpoint):
```

Comment:
If this is only used in tests, I think it's preferable to keep it in test code or utilities.

Reply:
Ok, I didn't check follow-on changes for usage.

If the goal is to allow locally overriding the API endpoint though, it seems better to me to just have `--api_endpoint` override whatever's returned from the frontend's ServerInfo endpoint, rather using it to construct a new ServerInfo without hitting the frontend at all.  As-is, it's semantically not really right to always use `/experiment/` as the URL pattern here, since that should be determined by the frontend.

Or concretely, if I were running local storzy but against the hosted-tensorboard-dev Spanner, I might want to be able to run `tensorboard dev --origin=https://hosted-tensorboard-dev.appspot.com --api_endpoint=localhost:10000` and I would argue that should use the experiment pattern and other metadata fetched from the origin still.

If it would seem weird if specifying a local `--api_endpoint` but not specifying `--origin` for it to default to pinging `tensorboard.dev`, we could change the default for `--origin` so that it only defaults to `tensorboard.dev` when `--api_endpoint` is not passed, and otherwise it defaults to empty and then only in that case do we locally construct the `ServerInfo` (and maybe in that case we should just use a dummy experiment URL pattern).

Reply:
Agreed; I also wasn’t fully content with the flags formulation in #2879.
Your approach sounds fine to me: I’ll revise that PR accordingly.